### ISSUE TITLE
Incorrect rendering using transform_matrix for novel views in 3DGS

### ISSUE THREAD

**Issue by Guochh01** [DL3DV-10K/Dataset#44]
Hello,

I'm trying to render novel-view images using a trained 3DGS model (without using COLMAP cache).For the new viewpoints, I used the `transform_matrix` parameter from each scene's `transforms.json` as the camera extrinsic.

I noticed that `transform_matrix` is in the OpenGL coordinate system(@ShengCN in https://github.com/DL3DV-10K/Dataset/issues/4#issuecomment-2016987462), and I followed the pose transformation suggested by @thucz in https://github.com/DL3DV-10K/Dataset/issues/4#issuecomment-2545016805:

```python
c2w = np.array(frame_data["transform_matrix"])
# flip z axis in world space
c2w[2, :] *= -1
# switch y and z axis in world space.
c2w = c2w[np.array([0, 2, 1, 3]), :]
# flip y and z axis in camera space. 
c2w[0:3, 1:3] *= -1
```

However, the rendered results are misaligned/distorted.

<div align="center">
  <img src="https://github.com/user-attachments/assets/9b579cc3-d723-442d-9fc2-e50e63ccc0ef" width="515" height="299" alt="Image" />
</div>

Additionally, upon reviewing the `colmap_utils.py` script from Nerfstudio, I found that the pose transformation might be influenced by the `keep_original_world_coordinate` parameter. Specifically, [lines 438–470](https://github.com/nerfstudio-project/nerfstudio/blob/0e889f7cb1c81681dd16b1e618386902ae44434d/nerfstudio/process_data/colmap_utils.py#L438-L470) in the script suggest that when `keep_original_world_coordinate` is set to `True`, the world coordinate system is preserved, which could affect the transformation process.

As also mentioned in [nerfstudio-project/nerfstudio#1504 (comment)](https://github.com/nerfstudio-project/nerfstudio/issues/1504#issuecomment-2194819450), the OpenGL-to-COLMAP transformation should only require flipping the directions of the Y and Z axes.

Could you please clarify:

- Is any additional transformation needed before using these matrices for novel view rendering (e.g., the `applied_transform` in `transforms.json`)?

- Does the `keep_original_world_coordinate` parameter in Nerfstudio's `colmap_utils.py` affect the pose transformation process?

Any guidance would be greatly appreciated!

**Comment by ShengCN** [comment]
Usually we use this script to convert the format + image undistortion using this script: https://github.com/graphdeco-inria/gaussian-splatting/blob/main/convert.py for nerfstudio generated camera transforms matrices.

**Comment by Guochh01** [comment]
Hello @ShengCN, I think I didn’t explain clearly before. I’ve only trained 3DGS (using gsplat) with part of the images from each scene, and I’d like to use the `transforms.json` in the dataset to provide extrinsics for other viewpoints.

As for the `convert.py` script you mentioned, it doesn’t seem to involve the `transforms.json` file — it just passes the image data into COLMAP. So I’m not sure what you meant there.

Thanks for clarifying!

**Comment by ShengCN** [comment]
I see your point now. Which codebase are you using? Usually you can directly use the original [3D GS code](https://github.com/graphdeco-inria/gaussian-splatting) to train on DL3DV scenes. Maybe you can first, have a try on 3D GS codebase and test if it looks correct. If it is OK, then the transform.json looks good. Next step is to read the dataloader and see how 3D GS's transformation matrix is different with your implementation.

**Comment by Guochh01** [comment]
Thank you for your reply! @ShengCN 

I’m using the training code from [gsplat](https://github.com/nerfstudio-project/gsplat/blob/65042cc501d1cdbefaf1d6f61a9a47575eec8c71/examples/simple_trainer.py) and set the `normalize_world_space` parameter to `False`. From my understanding, I only need to convert the OpenGL coordinates in the `transforms.json` to COLMAP/OpenCV coordinates, and then they should work for rendering novel views.

**Comment by Guochh01** [comment]
Hello @ShengCN !
I tested with `frame_00080.png` and found something unexpected. When using the extrinsic information obtained from `convert.py` (via `images.bin`), I was able to render the image correctly. However, when using the extrinsics from the dataset’s `transforms.json`, the rendered result was incorrect.

Here is the code I used to process the `transform_matrix` from `transforms.json`:

```python
c2w = np.array(frame_data["transform_matrix"])
# flip z axis in world space
c2w[2, :] *= -1
# switch y and z axis in world space.
c2w = c2w[np.array([0, 2, 1, 3]), :]
# flip y and z axis in camera space. 
c2w[0:3, 1:3] *= -1

# get the world-to-camera transform and set R, T
w2c = np.linalg.inv(c2w)
R = np.transpose(w2c[:3,:3])  # R is stored transposed due to 'glm' in CUDA code
T = w2c[:3, 3]

print("R =\n", R)
print("T =", T)
```

Output:
```bash
R =
 [[-0.03553145  0.99446795  0.09884844]
 [ 0.44883517 -0.07249453  0.89066915]
 [ 0.8929079   0.07601342 -0.44377635]]
T = [-0.5579683  -1.10419597  3.75337515]
```

By contrast, when I read the extrinsics from `images.bin`, the result is different:

```python
from scene.dataset_readers import read_extrinsics_binary
from scene.colmap_loader import qvec2rotmat

cameras_extrinsic_file = os.path.join(scene_path, "undistorted/sparse/0", "images.bin")
cam_extrinsics = read_extrinsics_binary(cameras_extrinsic_file)

for idx, key in enumerate(cam_extrinsics):
    if idx == 79:  
        extr = cam_extrinsics[key]
        R = np.transpose(qvec2rotmat(extr.qvec))
        T = np.array(extr.tvec)
        print(f"Camera index: {idx}, COLMAP image_id: {key}")
        print("R =\n", R)
        print("T =", T)
        break
```

Output:
```bash
R =
 [[ 0.99343863 -0.01280109  0.11364781]
 [-0.00345358  0.98990502  0.14169026]
 [-0.11431433 -0.14115307  0.98336567]]
T = [-0.12779081 -0.87875687  2.22196909]
```
The two sets of extrinsics are clearly different.

Could you clarify why this discrepancy exists? Does it mean that the provided `transforms.json` extrinsics cannot be directly used for rendering?

**Comment by RuiHuang9** [comment]
Hello friend, may I ask how you solve the problem?
